tc/core/polyhedral/cuda/memory_promotion_heuristic.cc

```diff
@@ -323,11 +323,138 @@ bool promotionImprovesCoalescing(
+ * accessed elements is single-valued.
+ * TODO: note that if a group is formed from partially overlapping references,
+ * one must consider per-reference access relation for single-valuedness as
+ * different references may have different values, but all of them remain
+ * independent of non-unrolled loop iterators.
+ */
+bool accessSubscriptsAreUnrolledLoops(
+    const TensorReferenceGroup& group,
+    const detail::ScheduleTree* root,
+    const detail::ScheduleTree* scope,
+    isl::multi_union_pw_aff prefixSchedule,
+    isl::multi_union_pw_aff threadSchedule) {
+  using namespace detail;
+
+  auto nodes = ScheduleTree::collect(scope);
+  auto leaves = functional::Filter(
+      [](const ScheduleTree* tree) { return tree->numChildren() == 0; }, nodes);
+
+  auto domainNode = root->elemAs<detail::ScheduleTreeElemDomain>();
+  CHECK(domainNode);
+  auto domain = domainNode->domain_;
+
+  // Descend into every leaf.
+  for (auto leaf : leaves) {
+    auto ancestors = leaf->ancestors(root);
+    ancestors.push_back(leaf);
+    auto subdomain = activeDomainPointsBelow(root, leaf);
+
+    auto unrolledDims = isl::union_pw_aff_list(leaf->ctx_, 1);
+    for (auto node : ancestors) {
+      auto band = node->elemAs<detail::ScheduleTreeElemBand>();
+      if (!band) {
+        continue;
+      }
+
+      isl::multi_union_pw_aff schedule = band->mupa_;
+      if (isThreadMappedBand(node)) {
+        // Band members are mapped to threads in inverse order.  There may be
+        // less members than thread dimensions (the remaining thread dimensions
+        // are mapped to 0).  Transform the thread schedule accordingly before
+        // subtracting it from the band schedule.
+        auto bandCompatibleThreadSchedule =
+            isl::union_pw_aff_list(schedule.get_ctx(), band->nMember());
+        auto threadScheduleDims =
+            threadSchedule.get_union_pw_aff_list().reverse();
+        CHECK_LE(band->nMember(), static_cast<size_t>(threadScheduleDims.n()));
+        for (int i = 0; i < band->nMember(); ++i) {
+          bandCompatibleThreadSchedule =
+              bandCompatibleThreadSchedule.add(threadScheduleDims.get(i));
+        }
+
+        schedule = schedule.sub(isl::multi_union_pw_aff(
+            schedule.get_space(), bandCompatibleThreadSchedule));
+      }
+      schedule = schedule.intersect_domain(subdomain);
+
+      for (size_t i = 0, e = band->nMember(); i < e; ++i) {
+        if (!band->unroll_[i]) {
+          continue;
+        }
+        unrolledDims = unrolledDims.add(schedule.get_union_pw_aff(i));
+      }
+    }
+
+    auto space = isl::space(leaf->ctx_, 0, unrolledDims.n())
+                     .align_params(subdomain.get_space());
+    auto unrolledDimsMupa = isl::multi_union_pw_aff(space, unrolledDims);
+
+    // It is possible that no loops are unrolled, in which case
+    // unrolledDimsMupa is zero-dimensional and needs an explicit domain
+    // to be convertible to a union_map.
```

Comment:
Did you actually run into this problem?
You're concatenating with the other schedules first, so I wouldn't expect this to be an issue.

Reply:
I did, hence the explanatory comment, although it may have been at some previous version of the code.

Reply:
On Wed, Jun 13, 2018 at 11:15:48AM -0700, ftynse wrote:
> I did, hence the explanatory comment, although it may have been at some previous version
 of the code.

Hmm... I guess it doesn't hurt, but it may point to a bug in isl,
in which case we should fix that instead of working around it.